### PR TITLE
Implementation of leaderboard data

### PR DESCRIPTION
**Pull request by AliNasra**
I have implemented:
1- updateBadges in ProfileDB:
Updates the badge table based on the updated scores in the has_domain_point table
2- getRankPerTag in ProfileDB:
Retrieves the top 100 scoring users given a topic string input. Associated service functions and endpoints with their swagger documentation are implemented!
The new endpoint is /leaderboard/{topic}

### CODE REVIEW

File: prediction-polls/backend/src/repositories/ProfileDB.js [hunk 1]
```diff
@@ -282,5 +282,30 @@ async function followerProfiles(userId) {
         return { error: errorCodes.DATABASE_ERROR };
     }
 }
+async function getRankPerTag(topic){
+    try {
+        const query = 'SELECT user.id , user.username has_domain_point.amount from user, has_domain_point where user.id = has_domain_point.userID AND has_domain_point.topic = ? Order By has_domain_point.amount Desc LIMIT 100';
```
Comment (EmreBatuhan) [inline prediction-polls/backend/src/repositories/ProfileDB.js:287]: I think the sql is wrong. could u revise it?

File: prediction-polls/backend/src/repositories/AuthorizationDB.js [hunk 2]
```diff
@@ -252,6 +252,9 @@ async function decrementUserParticipate(userId){
         throw error;
     }
 }
+
```
Comment (EmreBatuhan) [inline prediction-polls/backend/src/repositories/AuthorizationDB.js:255]: Don't just push empty rows

File: prediction-polls/backend/src/repositories/ProfileDB.js [hunk 3]
```diff
@@ -282,5 +282,30 @@ async function followerProfiles(userId) {
...
+        const [result] = await pool.query(query, [topic]);
+        return {result: result};
+    } catch (error) {
+        return { error: errorCodes.DATABASE_ERROR };
+    }
+}
+
+async function updateBadges(userId){
+    try {
+        const query = 'SELECT userId, topic, RANK() OVER (PARTITION BY topic ORDER BY amount DESC) as userRank FROM has_domain_point where userRank > 4';
```
Comment (EmreBatuhan) [inline prediction-polls/backend/src/repositories/ProfileDB.js:297]: This sql also has some problems. Please fix it also

File: prediction-polls/backend/src/repositories/ProfileDB.js (hunk 1 above)
Comment (AliNasra) [reply prediction-polls/backend/src/repositories/ProfileDB.js]: Fixed it
SELECT users.id , users.username, has_domain_point.amount from users, has_domain_point where users.id = has_domain_point.userId AND has_domain_point.topic = ? Order By has_domain_point.amount Desc LIMIT 100

File: prediction-polls/backend/src/repositories/ProfileDB.js (hunk 3 above)
Comment (AliNasra) [reply prediction-polls/backend/src/repositories/ProfileDB.js:297]: Select * from (SELECT userId, topic, RANK() OVER (PARTITION BY topic ORDER BY amount DESC) as userRank FROM has_domain_point) as temp where temp.userRank > 4
image

File: prediction-polls/backend/src/routes/ProfileRouter.js [hunk 4]
```diff
@@ -477,4 +477,49 @@ router.post('/followed',authenticator.authorizeAccessToken,service.getFollowedPr
  */
 router.post('/follower',authenticator.authorizeAccessToken,service.getFollowerProfiles)
 
+/**
+ * @swagger
+ * /leaderboard/{topic}:
```
Comment (EmreBatuhan) [inline prediction-polls/backend/src/routes/ProfileRouter.js]: It should be profiles/leaderboard/{topic}

**Review by EmreBatuhan — APPROVED** [review]
Looks good

**Pull request merged**